lib/client/signal_fx_client.js

```diff
@@ -19,26 +19,49 @@ var version = require('../../package.json').version;
  *
  * @constructor
  * @param apiToken
- * @param ingestEndpoint
- * @param apiEndpoint
- * @param timeout
- * @param batchSize
- * @param userAgents
+ * @param options - {
+ *                    enableAmazonUniqueId: boolean, // "false by default"
+ *                    dimensions:"object", // dimensions for each datapoint and event
+ *                    ingestEndpoint:"string",
+ *                    apiEndpoint:"string",
+ *                    timeout:"number",
+ *                    batchSize:"number",
+ *                    userAgents:"array"
+ *                }
  */
-function SignalFxClient(apiToken, ingestEndpoint,
-                        apiEndpoint, timeout,
-                        batchSize, userAgents) {
+function SignalFxClient(apiToken, options) {
+  var _this = this;
+
   this.apiToken = apiToken;
-  this.ingestEndpoint = ingestEndpoint || conf.DEFAULT_INGEST_ENDPOINT;
-  this.apiEndpoint = apiEndpoint || conf.DEFAULT_API_ENDPOINT;
-  this.timeout = timeout || conf.DEFAULT_TIMEOUT;
-  this.batchSize = Math.max(1, (batchSize ? batchSize : conf.DEFAULT_BATCH_SIZE));
-  this.userAgents = userAgents || null;
+  var params = options || {};
+
+  this.ingestEndpoint = params.ingestEndpoint || conf.DEFAULT_INGEST_ENDPOINT;
+  this.apiEndpoint = params.apiEndpoint || conf.DEFAULT_API_ENDPOINT;
+  this.timeout = params.timeout || conf.DEFAULT_TIMEOUT;
+  this.batchSize = Math.max(1, (params.batchSize ? params.batchSize : conf.DEFAULT_BATCH_SIZE));
+  this.userAgents = params.userAgents || null;
+  this.globalDimensions = params.dimensions || {};
+  this.enableAmazonUniqueId = params.enableAmazonUniqueId || false;
 
   this.queue = [];
   this.asyncRunning = false;
+
+  if (this.enableAmazonUniqueId) {
+    this._retrieveAWSUniqueId(function (isSuccess, AWSUniqueId) {
```

Comment:
How is this (or is this) protected from a race condition in which SignalFxClient is instantiated and then immediately .send is called to send datapoint while the unique id is still being retrieved?